apiserver/controllers/controllers.go

```diff
@@ -82,6 +94,22 @@ func handleError(w http.ResponseWriter, err error) {
 	json.NewEncoder(w).Encode(apiErr)
 }
 
+// GetControllerInfo returns means to identify this very garm instance.
+// This is very useful for debugging and monitoring purposes.
+func (a *APIController) GetControllerInfo() (hostname, controllerId string) {
```

Comment:
this func is supposed to be a nice helper, as I probably need hostname and controller id in all metrics (also future ones). I thought it would be clumsy to gather those values in all particulare places where I want to increment some metrics counter.
Example handleWorkflowJobEvent: this func needs hostname and controller id only for annotating the metrics, so I wanted to add as little as possible code to get those details, I hope this could help keep good readability?!

Reply:
Then how about this:

  * We extend: https://github.com/cloudbase/garm/blob/a0670c650352037bdabd5a53a0371fa21620d8b2/params/params.go#L276-L278 to include the controller hostname.
  * We create a function in ```runner.Runner{}``` that exposes the controller info, with the usual context validation, making sure whomever retrieves that info has access to do so. If we need that info internally, we can always use an admin context.
    * Checking authorization in all ```runner.Runner{}``` functions will help us in the future as well if we decide to split up components. In a zero trust environment (any remote system should be considered adversarial to be hones) we need to check if the remote **authenticated** component is authorized to perform a specific action. Right now, in most places we check for ```admin```, but with minimal planning, we can extend this to do much more complex validation in the future.
  * We populate the new hostname field with a call to ```os.Hostname()``` in ```runner.ControllerInfo(ctx context.Context)```
  * The ```runner.Runner{}``` type should have the ```controllerID``` cached. This way we don't need to hit the DB for a value that should never change. All we need is the hostname.
  * Use the ```runner.Runner{}``` we inject in the API controller (should already be there)/metrics collector to fetch that info.

This should keep concerns separated and readability at an acceptable level. 

(side note: If we really want to, we can define interfaces with a more narrow scope, that the ```runner.Runner{}``` type implements, and pass that along as a dependency to whichever components need it.)

I would like to avoid crossing boundaries.

  * The API controller should be used only as a presentation layer to the API consumer. It should not be used as a dependency in anything else.
  * The ```runner.Runner{}``` should be used as a central controller the API, the pool manager and any other component that needs to fetch info from the store or fetch jobs it needs to perform, calls into. This **should** be the only layer with direct access to the store.
  * The pool manager is the layer that **should** call into the ```runner.Runner{}``` to get the jobs it needs to perform to make sure that what we need is applied to the providers.
  * The providers are the glue between garm and the IaaS.

I realize this is not the case at the moment, and that we *do* cross boundaries right now, but I would like to avoid doing that more than necessary. The less we have to refactor later, the better.

The plan is to split these components up at some point, which will allow us to scale better if we need to. At that point we can replace the current dependencies we inject with a compatible ```RPC``` implementation that exposes the same functionality. This will hopefully allow us to not break everything when we split things up.

So it would be really helpful to group logic as well as possible and not spread it across multiple boundaries. I've seen this happen in the past, and it leads to spaghetti nightmare.

Does this sound sensible? Any suggestions are welcome.